USWDS-Site - Changelog: Create mockup with real content

amyleadem opened the pull request:
## Summary
Gathered changelog data for the banner component to stress test the current data structure and changelog presentation.

This PR also contains the most recent guidance for the proposed data structure for on-page changelogs. 

## Related issue
Closes #1776 

## Preview link
Preview link: [Banner changelog](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-changelog-mockup/components/banner/#banner-changelog)

## Related PRs
This PR builds on previous changelog PRs. They are available here for reference:
- USWDS-Site Changelog prototype https://github.com/uswds/uswds-site/pull/1753
- Changelog.md template https://github.com/uswds/uswds-team/pull/207

## Problem statement
The proposed changelog structure needs to be stress tested with real data to confirm that the structure will be usable and provides the optimal user experience. 

## Solution
Added real-world data to the banner component, dating back to release 2.7.0. 

### Methods for gathering data
The following tasks were performed to gather relevant change items:

- Searched USWDS release notes for the word "banner".
- Read through release notes to see if there are items related to banner, but not explicitly called out as a banner item (ex: uswds-init.js items)
- Checked history of component markdown files and added meaningful items
- Collected all found change items from 2.7.0 and above

### Other tasks completed in this PR
- Updated the naming convention for changelog data keys 
- Updated PR descriptions to include release notes/instructions for handling breaking changes, when not explicitly stated (ex: summary section in https://github.com/uswds/uswds/pull/4695)

### Outstanding questions:
- How far back do we go? (Note: This should be something we determine when generating the rollout plan in https://github.com/uswds/uswds-site/issues/1830) 

### Recommendations:
- Update guidelines for future PRs (Add more detail to https://github.com/uswds/uswds/pull/4842):
    - All PRs for both repos should have the component name in the title.
    - All USWDS PRs should explicitly state if the change is breaking and if so, provide clear instructions for user actions to remediate.
    
### Next tasks: 
- Determine rollout plan for changelogs (Issue #1830)
- Decide which data items are needed for collected changelogs (#1787)
- Demonstrate  the transition from USWDS version number to component package number (#1791)

## About on-page changelogs
### Proposed data structure
The information included in the changelogs will differ according to page type. Component pages will have more robust information with references to release number and breaking changes. Guidance pages will have show a simpler presentation.

#### Component page types:
**Demo of on-page display:**

![image](https://user-images.githubusercontent.com/93996430/198108609-f8267398-bfcc-4f03-83ee-efed72a65205.png)

**Proposed data structure:**
```
title:
type:
changelogURL:
items:
  - date:
    summary:
    additional_info:
    isBreaking:
    isCodeUpdate:
    pr:
    repo:
    version:
```

| Key | Description | Optional | Value type | Standard values | Displayed |
| :------ | ------ | ------ | ------ | ------ | ------ |
| title | Page title | no | string | N/A | No |
| type | Identify page type | no | string | Component, Documentation, Pattern, Template, Token, Utility | No |
| changelogURL |  url for CHANGELOG.MD | yes | string | N/A | Yes |
| date | Merge date | no | date | N/A | Yes |
| summary | Short description of change | no | string | N/A | Yes |
| additional_info | Additional description | yes | string | N/A | Yes |
| isBreaking | Breaking change | no | string | 'Yes', 'No', 'Potential' | Yes |
| isCodeUpdate | Change type | no | boolean | true, false | No |
| pr | Pull request number | no | number | N/A | Yes |
| repo | Pull request repo name | no | string | uswds, uswds-site | Yes |
| version | USWDS Version Number | no | number | N/A | Yes |

#### Guidance page types:

**Demo of on-page display:**
![image](https://user-images.githubusercontent.com/93996430/198108333-1f23a600-6f34-48b6-8d2c-600f0582fd72.png)

**Proposed data structure:**
```
title:
type:
items:
  - date:
    summary:
    additional_info:
    pr:
    repo:
```

| Key | Description | Optional | Value type | Standard values | Displayed |
| :------ | ------ | ------ | ------ | ------ | ------ |
| title |  Page title | no | string | N/A | No |
| type | Identify page type | no | string | Component, Documentation, Pattern, Template, Token, Utility | No |
| date | Merge date | no | date | N/A | Yes |
| summary | Short description of change | no | string | N/A | Yes |
| additional_info | Additional description | yes | string | N/A | Yes |
| pr | Pull request number | no | number | N/A | Yes |
| repo | Pull request repo name | no | string | uswds, uswds-site | Yes |


#### Collected changelogs
Collected changelogs should allow us to collate data into different presentations. These presentations will be decided with Issue https://github.com/uswds/uswds-site/issues/1787. Initial draft of design:

![image](https://user-images.githubusercontent.com/93996430/198108091-2efd77cc-ab16-4113-a9df-fbe942d33ca5.png)

amyleadem commented:
I have added a first pass at the banner changelog mockup, and was hoping to get early feedback on UX and data construction before I get too far down the road. 

In this PR, I have added real data for every change item I could find for banner, as far back as 2.7.0. Choosing 2.7.0 has no inherent meaning, it just felt like a good point to check in and ask how far back we wanted to go. 

Additionally, I have updated the data structure here according to the notes recorded in [issue 1787](https://github.com/uswds/uswds-site/issues/1787#issuecomment-1262729869). I imagine that initial review of this PR might lead to changes in the data structure and that these topics will continue to evolve. 

So at this initial phase, I was hoping to get feedback on if this structure is providing the optimal user experience and get guidance on how far back we should be going for data collection. 

If we affirm that this data structure and presentation meet our needs, I will go back through and confirm the data for every change item.

jaclinec commented:
@amyleadem This looks great to me - seems clear and laid out in a way that is easily scannable with the right amount of detail. I had one question. Are we still including a column for 'Type' of change - i.e. code or guidance? I did not see that in the preview.

amyleadem commented on _data/changelogs/accordion.yml:
```diff
@@ -0,0 +1,41 @@
+title: Accordion
+pageURL: components/accordion
```
Change pageURL - remove pathing when possible

amyleadem commented:
@thisisdano @mejiaj @mahoneycm 
I have made the following updates to the data structure. 
- Flattened the data structure
    - Created `summary` and `moreSummary` fields. (Open to suggestions to rename `moreSummary`)
- Replaced `title` with `pageTitle` for clarity.
- Removed `pageURL` and added `pageType` in an effort to reduce redundant data and the reliance on explicit pathing.
    - Example: if `pageTitle: Accordion` and `pageType: Component`, we should be able to build a path to `components/accordion` if we ever need it. 
    - `pageType` will also allow us to create more specific collections of changelog data if we ever desire it
    - Note: It will be trickier to create paths to nested pages like: https://designsystem.digital.gov/documentation/getting-started/developers/phase-one-install/, but I am thinking we can just provide explicit paths only for these more rare occurrences if needed. 
    - Open to suggestions for the Standard Values on this field. 
- Created `component-template.yml`, which contains inline comments that can be the basis for our data files.
    - If we like this inline comment option, I can either rework or remove the changelog-readme file.

mejiaj commented on _data/changelogs/accessibility.yml:
```diff
@@ -1,9 +1,8 @@
-    description:
-      summary: Updated all content to better explain current standards.
-      additional: This update explained USWDS accessibility standards, provided guidance on how to reach accessibility compliance on projects, and provided links to helpful resources.
+    summary: Updated all content to better explain current standards.
+    moreSummary: This update explained USWDS accessibility standards, provided guidance on how to reach accessibility compliance on projects, and provided links to helpful resources.
```
I feel like we were close before, but just needed small tweaks.

```yml
description: Updated all content to better explain current standards.
additional_info: |-
  This update explained USWDS accessibility standards, provided guidance on how to reach accessibility compliance on projects, and provided links to helpful resources.
```

amyleadem replied:
This is better. Thanks! Updated in [1d1a671](https://github.com/uswds/uswds-site/pull/1789/commits/1d1a67158cdfc137443d03ddb254ee6a90b061e3)

mejiaj commented on _data/changelogs/accessibility.yml:
```diff
@@ -1,9 +1,8 @@
-title: Accessibility
-pageURL: accessibility
+pageName: Accessibility
+pageType: Guidance
```
**Optional** we could simplify to `name/type`, but I can see the added clarity the page keyword brings

amyleadem replied:
I support it. There were initially more items up top so I liked the added clarity, but it doesn't seem necessary anymore. Updated in [51a9384](https://github.com/uswds/uswds-site/pull/1789/commits/51a9384c2feb9d225a743dbdac1d06d5d55053d7)

mejiaj commented on _data/changelogs/accordion.yml:
```diff
@@ -0,0 +1,36 @@
+  - date: 2022-08-05
+    summary: Added <code>type="button"</code> to the component code and guidance.
+    additional_info:
+    isBreaking: 'Potential'
+    isCodeUpdate: true
```
It looks like these aren't used on the accordion component page, is that intentional? It would be nice to see this information.

amyleadem replied:
I am not sure I understand. I noticed that I had forgotten to update the `changelog-table.html` file to reflect some of the new data keys. It should be displaying all the data now. Is this what you are referring to?

mejiaj replied:
Okay, I just double-checked and am seeing them now.

![image](https://user-images.githubusercontent.com/3385219/198144719-d6e80121-233b-491b-ab48-1913e061a60b.png)

mejiaj reviewed (approved):
Looks good, thanks @amyleadem!

amyleadem commented:
@thisisdano 
cc: @mejiaj 
Per our discussion yesterday, here are some proposed additional data fields:
```diff
title:
type:
changelogURL:
items:
  - date:
    summary:
    additional_info:
-   isCodeUpdate:
+   affectsAccessibility:
+   affectsAssets:
+   affectsGuidance:
+   affectsJavascript:
+   affectsMarkup:
+   affectsStyles:
    isBreaking:
    pr:
    repo:
    version:
```

The list is somewhat lengthy, but these switches would only needed to be included in the data if their value is `true`.
The additional fields will give us flexibility in the future to not only filter by change type, but combining this with the `isBreaking` field can give us enhanced flexibility for letting users know if additional action is required. 
For example:
- If `isBreaking` is true, we can choose to inform users that an update is `Required`
- If `affectsMarkup` or `affectsAssets` is true and `isBreaking` is false, we can inform users that an update is `Recommended`
- If `affectsStyles` or `affectsJavascript` is true and `isBreaking` is false, we can inform users that no additional action is needed.
- Additionally, `affectsAccessibility` can be used as a flag for mandatory action. So if `affectsAccessibility` and `affectsMarkup` are both true and `isBreaking` is false, we could still choose to let users know that action is `Required`.

Let me know if this is in line with your thinking and if you have any suggestions or questions. 

Once we decide on our keys, I will update the files and docs.

mejiaj commented on _data/changelogs/_CHANGELOG-README.md:
```diff
@@ -0,0 +1,45 @@
+  This is a guide for creating data for the "Latest Updates" section. Some things to remember:
+  - Tab spacing matters in .yml files. Incorrect spacing can cause errors.
+  - All comments here are instructional only and should be deleted if copied into the changelog .yml file. -->
+title: [Add page title. Ex: Accordion, Accessibility]
+type: [Add page type. Standard values: Component, Documentation, Pattern, Template, Token, Utility]
+<!-- At this time, we do not have any packages with their own changelog files, so this item can be ignored. -->
+changelogURL:[Provide url for the package CHANGELOG.md.]
+<!-- Add a list of changelog items in reverse chronological order-->
+items:
+  <!-- Start new data item -->
+  - date: [Add item release date in YYYY-MM-DD format.]
+    <!--
+    Add a concise summary of the change.
+      This summary can often be pulled from the related release notes or PR summary and adjusted to fit the changelog format.
+      Changelog summaries should be written in the past tense, use plain language, and begin with a consistent set of verbs whenever possible:
+        Examples: "Fixed", "Added", "Improved", "Optimized", "Updated", "Introduced"
+    If a single release item accomplished multiple tasks, break the tasks up into separate data items.
+        The goal is to itemize and highlight user benefits and actions required.
+    If a change item requires user action, make sure that the related PR clearly states a summary of the action required.
+    -->
+    summary: [Add a short, clear sentence that summarizes the change.]
+    additional_info: [If needed, add more detail in 1-2 short sentences.]
+    isBreaking: [Indicate if user action is required to prevent the system from breaking. Standard values: true, false]
+    isCodeUpdate: [Indicate if this update includes code updates. Standard values: true, false]
+    <!-- Indicate where users can find more information -->
+    pr: [Add related pull request (PR) number. Ex: 4695]
+    repo: [Declare the GitHub repo that contains related PR. Standard values: uswds, uswds-site]
+    version:[Add related USWDS or package release number. Ex: 3.1.0. Guidance updates will likely not have a related release.]
```
We could increase clarity by separating out the `type` expected.

For example:

```yml
title: Add page title. _Ex: Accordion, Accessibility_ [type: "string"]
```

Or before the value

```yml
# Type: string
title: Add page title. _Ex: Accordion, Accessibility_ 
```

amyleadem replied:
I will work this info into the comments as well. Will push out more updates once we finalize the keys.

mejiaj reviewed (commented):
Added some suggestions.

Not sure if I'm understanding the need on the proposal for these keys.

```diff
+   affectsAccessibility:
+   affectsAssets:
+   affectsGuidance:
+   affectsJavascript:
+   affectsMarkup:
+   affectsStyles:
```

It _might_ be an easier experience for developers to group this?

```yml
affects:
  - accessibility
  - assets
  - guidance
     etc
```

And on the template we check the logic by looping over `affects` key.

amyleadem commented on _data/changelogs/_CHANGELOG-README.md:
```diff
@@ -0,0 +1,45 @@
+
+
+<!-- Empty data template -->
+title:
+type:
+changelogURL:
+items:
+  - date:
+    summary:
+    additional_info:
+    isBreaking:
+    isCodeUpdate:
+    pr:
+    repo:
+    version:
```
I have notes on which fields are optional in the PR description - I can work this info into the comments.

amyleadem commented:
@mejiaj 
From my understanding, the main motivation for adding these keys is that one of the main questions users might have when looking at a changelog is: "Do I need to take any additional action here?"

We identify breaking changes, but we don't really mark when things like markup or assets  _should_ or _must_ be changed (for reasons other than breaking - like accessibility). 

Initially, we considered a simple field like `userActionNeeded`, but realized that we could instead infer this data if we identified more change types and combined them with `isBreaking`. The `affects` fields also have the added benefit of providing us with useful data that will allow us to display and filter change items by type in collected changelogs. Additionally, it also could give us more flexibility to even redefine user action needs in the future. 

As far as timing, it felt simpler to identify these fields while we are doing initial data entry, rather than needing to cycle back through everything later. 

As far as data structure goes, I am not sure I see a difference in the actual dev ease of use in either structure, but I could just be thinking too narrowly. My only slight style lean is that it might reduce entry error if each item had its own line and was a simple boolean. Definitely open to a different structure, so I am happy to defer. 

I hope that answers your questions!

thisisdano commented:
I've updated the keys for better consistency across themes.

- Used lowercase values for `type` (`Component` → `component`) 
- Removed `isCodeUpdate` since it's redundant with new `affects` keys
- `additional_info` → `summaryAdditional`
- `pr` → `githubPr`
- `repo` → `githubRepo`
- `version` → `versionUswds` (there may be `versionPackage` in the future)
- Changed `isBreaking` to a bool for simplicity. Cases where we think there is a _potential_ breaking change should be treated as a breaking change (`true`).
- Added the following keys to indicate the scope or domain of the change. This is meant to help users understand what part of the code changed in each entry.
  - `affectsAccessibility`
  - `affectsMarkup`
  - `affectsJavascript`
  - `affectsStyles`
  - `affectsContent`
  - `affectsGuidance`
  - `affectsAssets`

| Key | Description | Optional | Value type | Standard values | Displayed |
| :------ | ------ | ------ | ------ | ------ | ------ |
| title | Page title | no | string | N/A | No |
| type | Identify page type | no | string | component, documentation, pattern, template, token, utility | No |
| changelogURL |  url for CHANGELOG.MD | yes | string | N/A | Yes |
| date | Merge date | no | date | N/A | Yes |
| summary | Short description of change | no | string | N/A | Yes |
| summaryAdditional | Additional description | yes | string | N/A | Yes |
| isBreaking | Breaking change | no | boolean | true, false | Yes |
| affectsAccessibility | Change affects accessibility | yes | boolean | true, false | yes |
| affectsMarkup | Change affects component markup | yes | boolean | true, false | yes |
| affectsJavascript | Change affects component JS | yes | boolean | true, false | yes |
| affectsStyles | Change affects component styles, appearance, or settings | yes | boolean | true, false | yes |
| affectsContent | Change affects content in the component itself | yes | boolean | true, false | yes |
| affectsGuidance | Change affects component guidance | yes | boolean | true, false | yes |
| affectsAssets | Change affects assets related to a component, like images | yes | boolean | true, false | yes |
| githubPr | Pull request number | no | number | N/A | Yes |
| githubRepo | Pull request repo name | no | string | uswds, uswds-site | Yes |
| versionUswds | USWDS Version Number | no | number | N/A | Yes |

thisisdano commented:
I've also made a couple presentational changes.

- Separated the github link from the description for legibility
- Added an `Affects` column
- Highlighted breaking changes with a red tag
- Highlighted accessibility changes with a gold tag

<img width="1186" alt="Screenshot 2022-11-01 at 11 43 21 AM" src="https://user-images.githubusercontent.com/11464021/199313343-6185bfa2-7145-4953-8a1b-5b65161ee922.png">

amyleadem commented:
@thisisdano I was able to get the site to build. It was getting confused about the value of `changelogURL`, but commenting out the blank data fields resolved the issue. 

Additionally, I removed the comments on the blank items in the component `.yml` files because after doing some preliminary data entry today, it was much simpler to move forward without the comments. I believe we can rely on `_component-template.yml` instead as our guide for entry. 

All in all, I think the data setup looks great and we should be ready to start migrating data. 

For the display, I did want to flag that the content in the table display gets a bit - wobbly - at tablet widths. I think we can probably clear that up with some min-widths or alternate display for these widths. I will look into solutions tomorrow. This shouldn't be a blocker for  migration.
![image](https://user-images.githubusercontent.com/93996430/199350049-5b754246-1fdb-4b29-b7d3-99366af1ea70.png)

amyleadem commented:
@thisisdano @mejiaj 
I committed a proposed solution for the table display.  [d38522c](https://github.com/uswds/uswds-site/pull/1789/commits/d38522c1f8e9a6029d75c8ffbf5c3152c424a497)

## Proposal:
This commit does the following:
- At medium widths:
    - Moves the `Breaking` flag to the description cell
    - Hides the date
- At any width:
    - Shrinks mono font from 2xs to 3xs
    - Replaces `.flex` classes with `.grid-col`


### Narrow:

![image](https://user-images.githubusercontent.com/93996430/199542520-8ef21d83-25be-4b33-89aa-48305b08bc1b.png)

### Medium:

![image](https://user-images.githubusercontent.com/93996430/199543843-db67a0ac-7dba-40ac-952f-d140673af892.png)

### Wide:

![image](https://user-images.githubusercontent.com/93996430/199543489-e39df562-f0d2-4d48-9bb5-e49ce7a48738.png)

Let me know if you have feedback!

amyleadem commented:
@thisisdano @mejiaj 
I merged this work into `al-changelog-main` to create a central hub into which we can merge all subsequent data and code revisions. From this point on, we should branch all changes off of (and merge into) `al-changelog-main`.

Additionally:
- Opened  issue #1869 to track the fix for the table presentation at tablet widths.
- Drafted PR https://github.com/uswds/uswds-site/pull/1870 to more easily view changes in `al-changelog-main`

Please let me know if you have any questions.
